### ISSUE TITLE
disallow `from_exposed_addr(ptr.addr())`

### ISSUE THREAD

**Issue by KisaragiEffective** [rust-lang/rust-clippy#10383]
### What it does

disallows `from_exposed_addr(ptr.addr())` if the code is under [`#![feature(strict_provenance)]`](https://github.com/rust-lang/rust/issues/95228). The argument of `from_exposed_addr` can be value of [`expose_addr`](https://doc.rust-lang.org/stable/std/primitive.pointer.html#method.expose_addr) while `addr` cannot.

### Lint Name

create_ptr_from_unexposed_addr

### Category

correctness

### Advantage

- Prevents UB.

### Drawbacks

None

### Example

```rust
#![feature(strict_provenance)]
fn main() {
    let elem = 0u64;
    let x: *const u8 = std::ptr::from_exposed_addr((&elem as *const u64).addr())
}
```

Could be written as:

```rust
    let elem = 0u64;
    let x: *const u8 = std::ptr::from_exposed_addr((&elem as *const u64).exposed_addr())
```

**Comment by Alexendoo** [comment]
I've read a bit about provenance but not super in depth on the proposed rust API, would this sort of thing come up? I can't see the reason why you would want to call `from_exposed_addr` when you already have the pointer in the first place

**Comment by briansmith** [comment]
https://github.com/rust-lang/rust/pull/122964 renamed `exposed_addr` to `exposed_provenance()`, https://doc.rust-lang.org/std/primitive.pointer.html#method.expose_provenance.
https://github.com/rust-lang/rust/pull/122935 renamed `from_exposed_addr` to `with_exposed_provenance`,, https://doc.rust-lang.org/std/ptr/fn.with_exposed_provenance.html. Both were stabilized in 1.84.

I guess such a lint should also warn against `x.addr() as usize` if `x` is not a temporary (if `x` is a temporary, then `as usize` causes temporary lifetime extension.
